### PR TITLE
Additional Source Term in the Surface Heat Balance Calculation

### PR DESCRIPTION
**Pull request by xuanluo113**
Pull request overview
---------------------
Some heat transfer processes can be complicated and not necessary the required weather parameters (e.g., rain, ice on roof) are available in the commonly used TMY weather files. Current implementation of EnergyPlus does not support importing the pre-calculated results of these heat transfer processes. The proposed new feature will allow an additional source term to be included in the surface heat balance calculation.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by nealkruis** [comment]
@XuanLuoLBL @hongtz68 I think we need a field to define which side of the surface the heat flux is applied to. See [SurfaceProperty:ConvectionCoefficients](http://bigladdersoftware.com/epx/docs/8-8/input-output-reference/group-advanced-surface-concepts.html#surfacepropertyconvectioncoefficients).

**Comment by xuanluo113** [comment]
@nealkruis Thanks for the comments.
The original assumption is the Outside surface, but we agreed to add a field to make it more generic.
```
  A2 , \field Heat Source Location
       \required-field
       \type choice
       \key Outside
       \key Inside
       \default Outside
       \note This field indicates where the heat source applies.

```

**Comment by nealkruis** [comment]
Thanks @XuanLuoLBL!

**Comment by Myoldmopar** [comment]
```
\memo For Advanced/Research Usage
```

No. :-1: 

```
SurfaceProperty:HeatSource,
```

I don't really like this name.  It seems like this is defining a physical source of heat in the surface like a radiant system.  If we can't come up with anything better, then it's OK, but I'd like it to describe more of what it is doing to the heat balance (differential equation).  Maybe:

```
SurfaceProperty:HeatBalanceSourceTerm
SurfaceProperty:HeatBalanceSourceTermModifier
SurfaceProperty:HeatBalanceForcingFunction
SurfaceProperty:HeatBalanceSourceTermGain
SurfaceProperty:HeatBalanceAdditionalSourceTermValue
SurfaceProperty:DifferentialEquationForHeatBalanceSourceTerm
SurfaceProperty:TemperatureIndependentSourceTermForHeatBalance
```
OK, the last ones aren't for real, just brainstorming.

**Comment by nealkruis** [comment]
I agree with @Myoldmopar. Continuing the brainstorm:

```
SurfaceProperty:HeatFlux
SurfaceProperty:HeatFluxTerm
SurfaceProperty:HeatFluxBoundary
SurfaceProperty:HeatFluxBoundaryTerm
SurfaceProperty:HeatFluxForcingFunction
SurfaceProperty:HeatFluxBoundaryForcingFunctionTermYouKnowForRainNStuff
```
"Heat Flux" is the common term used for these kind of boundary conditions. That's what I lean towards.

**Comment by nrel-bot-3** [comment]
@XuanLuoLBL @lgentile it has been 14 days since this pull request was last updated.

**Comment by xuanluo113** [comment]
@Myoldmopar The feature is ready to be reviewed. We'd appreciate your help.

**Comment by Myoldmopar** [comment]
OK thanks @XuanLuoLBL.  For the record, the CI failures on GCC are unrelated and not to be concerned about for this.  I'll review it tomorrow.

File: idd/Energy+.idd.in
Comment (Myoldmopar) [inline idd/Energy+.idd.in]: I do not like this.  It's either suitable for use by all users or it doesn't go in the program.  This line can go away.

File: idd/Energy+.idd.in
Comment (Myoldmopar) [inline idd/Energy+.idd.in]: where the heat source **is applied**.

File: src/EnergyPlus/DataHeatBalSurface.cc
Comment (Myoldmopar) [inline src/EnergyPlus/DataHeatBalSurface.cc:227]: It would be awesome if these were just scalar member variables on the Surface data structure, instead of having to maintain yet two more arrays hanging out in global state needing to be allocated and deallocated.  I know this is a pattern that is already there, but if there are any significant changes to be made after this review, it would be great to have this become a new standard for future surface related enhancements.

File: src/EnergyPlus/DataSurfaces.hh
Comment (Myoldmopar) [inline src/EnergyPlus/DataSurfaces.hh]: This is not the most straightforward variable naming scheme.  Saying False to indicate that there is still some impact, just on the other side, is confusing to debug.  And what if the user applies a source to both sides??

How about instead of HasAdditionalHeatSourceTerm and OutsideAddHeatSourceTerm, you just use: HasOutsideHeatSourceTerm and HasInsideHeatSourceTerm.

And on the schedule variable here, if the user applies a source term to both the inside and outside, do you also need to break that into: OutsideHeatSourceTermSchedule and InsideHeatSourceTermSchedule?

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: With my comment above, this double layer if block would reduce to just a single `if HasOutsideHeatSourceTerm` block.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: With my comment above, this double layer if block would reduce to just a single `if HasInsideHeatSourceTerm` block.

File: src/EnergyPlus/SurfaceGeometry.cc
Comment (Myoldmopar) [inline src/EnergyPlus/SurfaceGeometry.cc]: This looks good except you should move the OutsideAddHeatSourceTerm and AdditionalHeatSourceTermSch variable declarations to inside the loop since they aren't used outside of it or from one loop iteration to the next.

File: tst/EnergyPlus/unit/HeatBalanceSurfaceManager.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/HeatBalanceSurfaceManager.unit.cc:2509]: Yeah, this makes sense given the inputs above.

**Review by Myoldmopar — CHANGES_REQUESTED** [review]
I went ahead and reviewed tonight.  I have some changes requested, and maybe some discussion about it, but I will say overall it doesn't seem very far off.  Let me know what you think on my comments and we should be able to push this along without any major issues unless other reviewers spot something I missed.

File: idd/Energy+.idd.in
Comment (mjwitte) [reply idd/Energy+.idd.in]: The PR title and description only mention an exterior source term, but it looks like the implementation allows for either inside or outside? Please update these to match.

File: idd/Energy+.idd.in
Comment (mjwitte) [reply idd/Energy+.idd.in]: Suggest that the memo read "Allows an additional heat source term to be added to the inside or outside surface boundary."

Do you allow a source to be added to both the inside and outside of the same surface?  If not, then this should be explained in the memo.  If yes, then this input structure has validation problems, because there could be more than one object with the same name. Does your code check to make sure that there is not more than one object setting, say, an inside source on the same surface?  I'm not sure about the other surfaceproperty objects, but I think the surface names must be unique across all SurfaceProperty:ConvectionCoefficients objects.

An alternate input design would be
```
SurfaceProperty:HeatBalanceSourceTerm,
 A1 , \field Surface Name
A2 ; \field Outside Surface Heat Source Term Schedule Name
A3 ; \field Inside Surface Heat Source Term Schedule Name
```
This would allow getinput to check for unique surface names within this class. The schedule name fields would then be optional instead of required if this approach is taken.  If blank, then there is no source on that side.

Also-do you check and throw and error if this is applied to a surface that has a special boundary condition such as OtherSideCoefficients or OtherSideConditions? 

Does it work with an interzone surface (outside boundary = surface or zone)?

File: testfiles/AddHeatSourceTerm_RefBldgSmallOfficeNew2004.idf
Comment (mjwitte) [inline testfiles/AddHeatSourceTerm_RefBldgSmallOfficeNew2004.idf]: This does not match the file name.  I know the comment block below is copied from RefBldgSmallOfficeNew2004_Chicago but it's not the expected format. If you look at ExampleFiles.html, the last column "Description" is blank for all of the RefBldg files.  Please add a `Highlights` section describing the main reason for this example file.  See 5ZoneAirCooledConvCoef.idf.

@Myoldmopar It would be nice to move the Description to column 2 in ExampleFiles.html.

File: testfiles/CMakeLists.txt
Comment (mjwitte) [inline testfiles/CMakeLists.txt]: I would name this `SurfaceHeatSourceTerm_RefBldgSmallOfficeNew2004.idf`.

File: doc/input-output-reference/src/overview/group-advanced-surface-concepts.tex
Comment (mjwitte) [inline doc/input-output-reference/src/overview/group-advanced-surface-concepts.tex]: Does this section belong in this pull request?

File: doc/input-output-reference/src/overview/group-advanced-surface-concepts.tex
Comment (mjwitte) [inline doc/input-output-reference/src/overview/group-advanced-surface-concepts.tex]: Can this only be applied to exterior surfaces with an "outdoors" boundary condition?

File: doc/engineering-reference/src/surface-heat-balance-manager-processes/outside-surface-heat-balance.tex
Comment (mjwitte) [inline doc/engineering-reference/src/surface-heat-balance-manager-processes/outside-surface-heat-balance.tex]: Is there an equation in the Engineering Reference where this term should be added?

File: idd/Energy+.idd.in
Comment (mjwitte) [reply idd/Energy+.idd.in]: Or better might be "Outside Face Heat Source Term . . " and "Inside Face . . . " to match the terminology used elsewhere in EnergyPlus.

Also - is there a new output variable for these terms?  If not, we need them, suggested names:
Surface Inside Face Heat Source Gain Rate per Area [W/m2]
Surface Inside Face Heat Source Gain Rate [W]
similar for Surface Outside Face . ..

**Comment by xuanluo113** [comment]
@Myoldmopar @mjwitte 

Thanks for your great comments. Considering both of your concerns, I think there're two ways to handle the two side thing:

(1) If I define two separate fields HasInsideHeatSourceTerm and HasOutsideHeatSourceTerm, I don't think there should be a problem with assigning 2 SurfaceProperty:HeatBalanceSourceTerm objects on one surface. We use HasHeatSourceTerm at the beginning as we didn't concern heat source term to the inside surface as well - added based on review comments. I'm going to check why surface names must be unique across all SurfaceProperty:ConvectionCoefficients objects and if it applies to this case.

(2) Change the structure to:
```
SurfaceProperty:HeatBalanceSourceTerm,
  A1 , \field Surface Name
  A2 , \field Inside Additional Heat Source Term
        \type choice
        \key Yes
        \key No
  A3 , \field Outside Additional Heat Source Term
        \ Yes or No
  A4, \field Inside Surface Heat Source Term Schedule Name
  A5; \field Outside Surface Heat Source Term Schedule Name
```
I think it's very unlike to define both sides and I think this structure could be a little misleading, so I feel (1) will do. I'll work on that and check.

And Mike, I think defining a scalar array for these kinds of variables is a great idea and applies to most of the arrays like this to speed up. But I'm not sure how E+ generally does that and a key to search. If you have something in mind using scalar types could you please point me to that?

**Comment by Myoldmopar** [comment]
> And Mike, I think defining a scalar array for these kinds of variables is a great idea and applies to most of the arrays like this to speed up. But I'm not sure how E+ generally does that and a key to search. If you have something in mind using scalar types could you please point me to that?

@XuanLuoLBL I'm the one who mentioned using a scalar variable, so I think this comment was for me.  I wouldn't hold up this PR for this improvement, but for kicks, here is what I mean.

Right now, we have a "Surface" struct, which defines attributes about a surface:
```
struct SurfaceDataOrSomething {
  string surfaceName;
  int someValue;
}
```

An array of these structs is allocated to the number of surfaces during surface get input:
```
Array1D< SurfaceDataOrSomething > surfaces;
surfaces.allocate( numSurfaces );
```

Although I have my thoughts on that already, let's skip ahead.  Right now we also have separate array variables which are allocated for each surface:
```
Array1D< Real64 > surfaceVar;
Array1D< Real64 > surfaceVar2;
Array1D< Real64 > surfaceVar3;
surfaceVar.allocate( numSurfaces );
surfaceVar2.allocate( numSurfaces );
surfaceVar3.allocate( numSurfaces );
```

Then inside calculation code, we end up with thousands of these:
```
someParameter = surfaceVar( surfaceNum ) + surfaceVar2( surfaceNum ) + surfaceVar3( surfaceNum );
```

Why not just add `surfaceVar` to the original surface struct?  That struct is already being allocated for each surface.
```
struct SurfaceDataOrSomething {
  string surfaceName;
  int someValue;
  Real64 surfaceVar;
  Real64 surfaceVar2;
  Real64 surfaceVar3;
}
```

Then those separate arrays never have to be allocated and managed separately.  Inside the calculation code, compare this result to the line above:
```
auto & thisSurface = surfaces( surfaceNum );
someParameter = thisSurface.surfaceVar + thisSurface.surfaceVar2 + thisSurface.surfaceVar3;
```

**Comment by Myoldmopar** [comment]
@XuanLuoLBL what about this approach?

```
SurfaceProperty:HeatBalanceSourceTerm,
  A1, \field Surface Name
  A2, \field Inside Surface Heat Source Term Schedule Name
        \note The value of this schedule is the source term value for the inside face of this surface
        \note If this field is left blank, no inside surface source term will be applied
  A3; \field Outside Surface Heat Source Term Schedule Name
        \note The value of this schedule is the source term value for the outside face of this surface
        \note If this field is left blank, no outside surface source term will be applied
```

This reduces your IDD down to simply a surface name reference and two schedule inputs.  The extra flags for whether a source term is applied are completely unnecessary and add extra maintenance to your code.  The existence of the schedule name is a robust signal for whether the user wants to impose a source term on that particular face.

File: idd/Energy+.idd.in
Comment (xuanluo113) [reply idd/Energy+.idd.in]: @Myoldmopar @mjwitte 
The feature is supposed to allow either or both sides of a surface. And it is supposed to work with any surfaces in all conditions. So we
- Updated the IDD
- Modified the docs
- Checked for unique surface name
- Threw errors if OtherSideCoefficients

File: src/EnergyPlus/DataSurfaces.hh
Comment (xuanluo113) [reply src/EnergyPlus/DataSurfaces.hh]: We updated the variables to have only OutsideAddHeatSourceTermSch and InsideAddHeatSourceTermSch.

File: src/EnergyPlus/SurfaceGeometry.cc
Comment (xuanluo113) [reply src/EnergyPlus/SurfaceGeometry.cc]: We modified this according to the updated variables.

File: doc/input-output-reference/src/overview/group-advanced-surface-concepts.tex
Comment (xuanluo113) [reply doc/input-output-reference/src/overview/group-advanced-surface-concepts.tex]: Not really. We modified the docs.

File: idd/Energy+.idd.in
Comment (Myoldmopar) [inline idd/Energy+.idd.in:16137]: Thanks for updating all this, it looks much better.  It looks like a tab character got put in the memo line below on accident.  I'd say hold off from fixing it until you know if there will be any other final changes.  I'll review it again asap.

File: testfiles/AddHeatSourceTerm_RefBldgSmallOfficeNew2004.idf
Comment (xuanluo113) [reply testfiles/AddHeatSourceTerm_RefBldgSmallOfficeNew2004.idf]: The name was modified and we also added the highlight section.

File: idd/Energy+.idd.in
Comment (xuanluo113) [reply idd/Energy+.idd.in:16137]: Okay. I found that I missed two more comments to address - adding output vars and add an equation to the engineering ref. I'll make another commit after your 2nd round review then.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: It's OK to leave it like this, but for an order independent loop like this, it would be nice to simplify it using range-based looping.  You can also eliminate the != 0 check, since in C++, a nonzero value is truthy:

```
for ( auto & thisSurf : Surface ) {
  if ( thisSurf.InsideHeatSourceTermSchedule ) {
    QAdditionalHeatSourceInside( SurfNum ) = EnergyPlus::ScheduleManager::GetCurrentScheduleValue( thisSurf.InsideHeatSourceTermSchedule );
  }
}
```

Consider this a note for later unless you really want to make the change now.

**Review by Myoldmopar — APPROVED** [review]
This is looking much better.  I know you still have some changes to push up, but I'll mark this approved and say it's ready unless I see something in those final changes.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: I have to keep the index here to get QAdditionalHeatSourceInside( SurfNum ), but it's good to know the preferred structure. I've adopted the nonzero=truthy way.

**Comment by xuanluo113** [comment]
@Myoldmopar Done. Hopefully.

**Comment by Myoldmopar** [comment]
This came a long way, and looks great now.  Built and tested locally, all is well.  Played with the inputs, nothing broke, verified the energy changes are showing up in the surface temperatures.  Merging.

**Pull request merged**
